Add drop_dims/new_dims kwargs to map_blocks

mrocklin opened the pull request:
This allows `map_blocks` to support functions that create or destroy dimensions.  

Example
-------

```python
In [1]: import dask.array as da

In [2]: x = da.arange(16, chunks=(8,))

In [3]: y = x.map_blocks(lambda blk: blk.reshape((2, 2, 2)), chunks=(2, 2, 2), new_dims=[1, 2])

In [4]: y
Out[4]: dask.array<atop-3b..., shape=(4, 2, 2), dtype=None, chunksize=(2, 2, 2)>

In [5]: y.compute()
Out[5]:
array([[[ 0,  1],
        [ 2,  3]],

       [[ 4,  5],
        [ 6,  7]],

       [[ 8,  9],
        [10, 11]],

       [[12, 13],
        [14, 15]]])
```

cc @alimanfoo @shoyer

mrocklin commented:
I'm thinking of renaming `drop_dims/new_dims` to `drop_axis/new_axis`.  Thoughts @alimanfoo ?

shoyer commented:
Yes, those are better names.

On Sat, Nov 28, 2015 at 2:59 PM, Matthew Rocklin <notifications@github.com>
wrote:

> I'm thinking of renaming `drop_dims/new_dims` to `drop_axis/new_axis`.  Thoughts @alimanfoo ?
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/blaze/dask/pull/848#issuecomment-160344043

alimanfoo commented:
Fine by me.

On Saturday, 28 November 2015, Matthew Rocklin <notifications@github.com>
wrote:

> I'm thinking of renaming drop_dims/new_dims to drop_axis/new_axis.
> Thoughts @alimanfoo <https://github.com/alimanfoo> ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/blaze/dask/pull/848#issuecomment-160344043>.
>


-- 
Alistair Miles
Head of Epidemiological Informatics
Centre for Genomics and Global Health <http://cggh.org>
The Wellcome Trust Centre for Human Genetics
Roosevelt Drive
Oxford
OX3 7BN
United Kingdom
Web: http://purl.org/net/aliman
Email: alimanfoo@googlemail.com <alimanfoo@gmail.com>
Tel: +44 (0)1865 287721

mrocklin commented:
Changed `drop/new_dims` to `drop/new_axis` (@alimanfoo this API change will affect your code).

Added more explanation about mapping across unequal chunk sizes as per @alimanfoo 's suggestion

mrocklin commented:
Merging soon if no comments
